Remove SharedMemory::adopt

donny-dont opened the pull request:
#### 139355b62834b66aeb6f812dfbc05f33a2e7686e
<pre>
Remove SharedMemory::adopt
<a href="https://bugs.webkit.org/show_bug.cgi?id=256676">https://bugs.webkit.org/show_bug.cgi?id=256676</a>

Reviewed by Kimmo Kinnunen and Fujii Hironori.

Use the updated methods in `Win32Handle` to implement
`Data::tryCreateSharedMemory` without directly calling into
`DuplicateHandle`. This code was the only place calling
`SharedMemory::adopt` outside of `SharedMemory.cpp`. Move the contents
of it into `SharedMemory::map` and remove this Window&apos;s only method.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
(WebKit::SharedMemory::map):
(WebKit::SharedMemory::adopt): Deleted.

Canonical link: <a href="https://commits.webkit.org/265140@main">https://commits.webkit.org/265140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1608c5dae076eaf0a544feb92737a0e6a481f81

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11622 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9656 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10169 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12596 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10913 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12006 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8220 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16372 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12456 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9688 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8847 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2388 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13076 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9466 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
```diff
@@ -113,12 +113,13 @@ RefPtr<SharedMemory> Data::tryCreateSharedMemory() const
+    auto dupHandle = fileHandle.copy();
+    auto handle = SharedMemory::Handle { WTFMove(dupHandle), m_size };
+
+    (void) fileHandle.leak();
```
Looks pretty bad. m_fileMapping of MappedFileData should be Win32Handle. Then, we can simply copy the Win32Handle.

donny-dont replied:
👍

fujii commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
```diff
@@ -113,11 +113,10 @@ RefPtr<SharedMemory> Data::tryCreateSharedMemory() const
         return nullptr;
 
-    auto newHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();
-    if (!newHandle)
```
Is it OK to remove this check?

fujii commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
```diff
@@ -113,11 +113,14 @@ RefPtr<SharedMemory> Data::tryCreateSharedMemory() const
         return nullptr;
 
-    return SharedMemory::adopt(newHandle.leak(), m_size, SharedMemory::Protection::ReadOnly);
+    auto dupHandle = fileHandle.copy();
```
This code copies the handle twice. For `fileHandle` and `dupHandle`. Don't copy twice. And, check the handle after copying.

fujii commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
```diff
@@ -113,11 +113,14 @@ RefPtr<SharedMemory> Data::tryCreateSharedMemory() const
 
-    auto newHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();
-    if (!newHandle)
+    auto fileHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();
```
> auto fileHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();

This code copies the handle.

> auto& fileHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();

This code doesn't copy the handle.


> auto fileHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping().copy();

This code copies the handle only once.

kkinnunen-apple replied:
Yeah, exactly as fujii said.
We should remove `MachSendRight::copySendRight()` and `Win32Handle::copy()` to prevent this confusion.
So you could do something along the lines of:
```
  auto& fileData = std::get<FileSystem::MappedFileData>(*m_buffer);
  return SharedMemory::map({ fileData.fileMapping(), m_size }, SharedMemory::Protection::ReadOnly);
```

fujii commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
```diff
@@ -113,11 +113,11 @@ RefPtr<SharedMemory> Data::tryCreateSharedMemory() const
-    auto newHandle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();
-    if (!newHandle)
+    auto handle = std::get<FileSystem::MappedFileData>(*m_buffer).fileMapping();
+    if (!handle)
```
You renamed this variable. But, I think old name is better because it indicated a new handle.

donny-dont committed:
Remove SharedMemory::adopt
https://bugs.webkit.org/show_bug.cgi?id=256676

Reviewed by Kimmo Kinnunen and Fujii Hironori.

Use the updated methods in `Win32Handle` to implement
`Data::tryCreateSharedMemory` without directly calling into
`DuplicateHandle`. This code was the only place calling
`SharedMemory::adopt` outside of `SharedMemory.cpp`. Move the contents
of it into `SharedMemory::map` and remove this Window's only method.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
(WebKit::SharedMemory::map):
(WebKit::SharedMemory::adopt): Deleted.

Canonical link: https://commits.webkit.org/265140@main
